worldedit-core/src/main/java/com/fastasyncworldedit/core/history/changeset/AbstractChangeSet.java

```diff
@@ -372,25 +364,58 @@ public Future<?> addWriteTask(final Runnable writeTask, final boolean completeNo
                 } else {
                     int hash = t.getMessage().hashCode();
                     if (lastException.getAndSet(hash) != hash) {
-                        t.printStackTrace();
-                    }
-                }
-            } finally {
-                if (AbstractChangeSet.this.waitingCombined.decrementAndGet() <= 0) {
-                    synchronized (AbstractChangeSet.this.waitingAsync) {
-                        AbstractChangeSet.this.waitingAsync.notifyAll();
-                    }
-                    synchronized (AbstractChangeSet.this.waitingCombined) {
-                        AbstractChangeSet.this.waitingCombined.notifyAll();
+                        LOGGER.catching(t);
                     }
                 }
             }
         };
         if (completeNow) {
             wrappedTask.run();
-            return Futures.immediateCancelledFuture();
+            return Futures.immediateVoidFuture();
         } else {
-            return Fawe.instance().getQueueHandler().submit(wrappedTask);
+            CompletableFuture<?> task = new CompletableFuture<>();
+            queue.add(() -> {
+                wrappedTask.run();
+                task.complete(null);
+            });
+            // make sure changes are processed
+            triggerWorker();
+            return task;
+        }
+    }
+
+    private void triggerWorker() {
+        if (workerSemaphore.availablePermits() == 0) {
+            return; // fast path to avoid additional tasks: a worker is already draining the queue
+        }
+        // create a new worker to drain the current queue
+        Fawe.instance().getQueueHandler().submit(() -> drainQueue(false));
+    }
+
+    private void drainQueue(boolean ignoreRunningState) {
+        if (!workerSemaphore.tryAcquire()) {
+            if (ignoreRunningState) {
+                // ignoreRunningState means we want to block
+                // even if another thread is already draining
+                try {
+                    workerSemaphore.acquire();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+            } else {
+                return; // another thread is draining the queue already, ignore
+            }
+        }
+        try {
+            while (true) {
+                var next = queue.poll();
+                if (next == null) {
```

Comment:
Why do we not just have the above in the while loop?

Reply:
We do not really profit from threads competing for processing the tasks from the queue

Reply:
This doesn't solve that though, it's the same as if it were just in the while itself no?

Reply:
I didn't find a good way to pull it into the loop, but I also don't see any benefit from that - did I miss something?

Reply:
Define next outside the loop and `while ((next = queue.poll()) != null)` (unless I'm missing something). No particular benefit other than fewer lines really and avoiding a while(true)

Reply:
I see, then I just totally misunderstood what you meant, lol. Will update the code.